arangod/Aql/SubqueryEndExecutor.cpp

```diff
@@ -0,0 +1,182 @@
+////////////////////////////////////////////////////////////////////////////////
+/// DISCLAIMER
+///
+/// Copyright 2019 ArangoDB GmbH, Cologne, Germany
+///
+/// Licensed under the Apache License, Version 2.0 (the "License");
+/// you may not use this file except in compliance with the License.
+/// You may obtain a copy of the License at
+///
+///     http://www.apache.org/licenses/LICENSE-2.0
+///
+/// Unless required by applicable law or agreed to in writing, software
+/// distributed under the License is distributed on an "AS IS" BASIS,
+/// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+/// See the License for the specific language governing permissions and
+/// limitations under the License.
+///
+/// Copyright holder is ArangoDB GmbH, Cologne, Germany
+///
+/// @author Michael Hackstein
+/// @author Markus Pfeiffer
+////////////////////////////////////////////////////////////////////////////////
+
+#include "ExecutionBlock.h"
+#include "SubqueryEndExecutor.h"
+
+#include "Aql/OutputAqlItemRow.h"
+#include "Aql/SingleRowFetcher.h"
+
+#include <velocypack/Builder.h>
+#include <velocypack/velocypack-aliases.h>
+
+
+using namespace arangodb;
+using namespace arangodb::aql;
+
+SubqueryEndExecutorInfos::SubqueryEndExecutorInfos(
+    std::shared_ptr<std::unordered_set<RegisterId>> readableInputRegisters,
+    std::shared_ptr<std::unordered_set<RegisterId>> writeableOutputRegisters,
+    RegisterId nrInputRegisters, RegisterId nrOutputRegisters,
+    std::unordered_set<RegisterId> const& registersToClear,
+    std::unordered_set<RegisterId> registersToKeep,
+    transaction::Methods* trxPtr, RegisterId outReg)
+    : ExecutorInfos(readableInputRegisters, writeableOutputRegisters, nrInputRegisters,
+                    nrOutputRegisters, registersToClear, std::move(registersToKeep)),
+      _trxPtr(trxPtr), _outReg(outReg) { }
+
+SubqueryEndExecutorInfos::SubqueryEndExecutorInfos(SubqueryEndExecutorInfos&& other) = default;
+
+SubqueryEndExecutorInfos::~SubqueryEndExecutorInfos() = default;
+
+SubqueryEndExecutor::SubqueryEndExecutor(Fetcher& fetcher, SubqueryEndExecutorInfos& infos)
+    : _fetcher(fetcher),
+      _infos(infos),
+      _state(ACCUMULATE) {
+  resetAccumulator();
+}
+
+SubqueryEndExecutor::~SubqueryEndExecutor() = default;
+
+std::pair<ExecutionState, NoStats> SubqueryEndExecutor::produceRows(OutputAqlItemRow& output) {
+  ExecutionState state;
+  NoStats stats;
+
+  InputAqlItemRow inputRow = InputAqlItemRow{CreateInvalidInputRowHint()};
+  ShadowAqlItemRow shadowRow = ShadowAqlItemRow{CreateInvalidShadowRowHint{}};
+
+  while (!output.isFull()) {
+    switch (_state) {
+      case ACCUMULATE: {
+        std::tie(state, inputRow) = _fetcher.fetchRow();
+
+        if (state == ExecutionState::WAITING) {
+          TRI_ASSERT(!inputRow.isInitialized());
+          return {state, std::move(stats)};
```

Comment:
Please do not move the stats here.
This might disallow return value optimization cases and NoStats is not expensive ;)